### PR TITLE
FINERACT-1678: Performance improvement of Apply Charge to Overdue Loans COB Business Step

### PR DESCRIPTION
**Pull request by adamsaghy**
## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/cob/loan/ApplyChargeToOverdueLoansBusinessStep.java [hunk 1]
```diff
@@ -32,26 +31,21 @@
...
-        // TODO: this is very much not effective to get all overdue installments for each loan, a new method needs to be
-        // implemented for it
-        Map<Long, List<OverdueLoanScheduleData>> groupedOverdueData = overdueLoanScheduledInstallments.stream()
-                .collect(Collectors.groupingBy(OverdueLoanScheduleData::getLoanId));
-        for (Long loanId : groupedOverdueData.keySet()) {
-            loanWritePlatformService.applyOverdueChargesForLoan(input.getId(), groupedOverdueData.get(loanId));
-        }
-        return input;
+    public Loan execute(Loan loan) {
+        log.info("-------STARTED--------{}", LocalDateTime.now(ZoneId.systemDefault()));
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/cob/loan/ApplyChargeToOverdueLoansBusinessStep.java]: I guess these were left in accidentally.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java [hunk 2]
```diff
@@ -115,11 +114,11 @@ public void setup() {
         this.systemTimeZone = TimeZone.getTimeZone(Utils.TENANT_TIME_ZONE);
     }
 
-    @AfterEach
-    public void tearDown() {
-        GlobalConfigurationHelper.resetAllDefaultGlobalConfigurations(requestSpec, responseSpec);
-        GlobalConfigurationHelper.verifyAllDefaultGlobalConfigurations(requestSpec, responseSpec);
-    }
+    // @AfterEach
```
Comment (galovics) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java]: Accidental?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java [hunk 3]
```diff
@@ -1558,6 +1560,42 @@ public Collection<OverdueLoanScheduleData> retrieveAllLoansWithOverdueInstallmen
...
+        final boolean backdatePenalties = configurationDomainService.isBackdatePenaltiesEnabled();
+
+        for (LoanRepaymentScheduleInstallment installment : loan.getRepaymentScheduleInstallments()) {
+            if (installment.isObligationsMet() || installment.isRecalculatedInterestComponent()) {
+                continue;
+            }
+
+            if (installment.isOverdueOn(DateUtils.getBusinessLocalDate().plusDays(penaltyWaitPeriod))) {
+                if (!backdatePenalties
+                        && installment.getDueDate().compareTo(DateUtils.getBusinessLocalDate().plusDays(penaltyWaitPeriod + 1)) < 0) {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java]: Would be nice to extract this compareTo condition into a variable with a proper name to better understand what it means.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java [hunk 4]
```diff
@@ -1558,6 +1560,42 @@ public Collection<OverdueLoanScheduleData> retrieveAllLoansWithOverdueInstallmen
...
+                    continue;
+                }
+                Optional<Charge> penaltyCharge = loan.getLoanProduct().getLoanProductCharges().stream()
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java:1586]: Is there a way we could extract this magic number (the 9) into a constant or reuse it from somewhere?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java [hunk 5]
```diff
@@ -1558,6 +1560,42 @@ public Collection<OverdueLoanScheduleData> retrieveAllLoansWithOverdueInstallmen
...
+                        .filter((e) -> e.getChargeTimeType() == 9 && e.isLoanCharge()).findFirst();
+
+                if (penaltyCharge.isEmpty()) {
+                    continue;
+                }
+
+                list.add(new OverdueLoanScheduleData(loan.getId(), penaltyCharge.get().getId(),
+                        DateUtils.DEFAULT_DATE_FORMATTER.format(installment.getDueDate()), penaltyCharge.get().getAmount(),
+                        DateUtils.DEFAULT_DATE_FORMAT, "en_GB", installment.getPrincipalOutstanding(loan.getCurrency()).getAmount(),
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java]: en_GB, could we reuse this from the Locale class?

**Review by galovics — CHANGES_REQUESTED** [review]

**Comment by adamsaghy** [comment]
@galovics Yeah... i wanted this only in my repo... its half done...just wanted to see how the tests go... :D
I guess i should not work around midnight :D

**Review by galovics — APPROVED** [review]
LGTM

**Pull request merged**